// @sap/cds-dk/lib/init/template/auditlog/index.js
const { join } = require('path')
const ProjectReader = require('../../util/projectReader')
const { mergeYAML } = require('../../util/merge')
const { copyAndTrack, parseMd5File, writeMd5File } = require('../../util/md5Tracking')
const { srvNode, srvJava, auditlog } = require('../_merging/registry-mta')

module.exports = class AuditlogTemplate extends require('../templateBase') {
    constructor(projectPath, generator) {
        super(projectPath, generator, __dirname)
        this.projectReader = new ProjectReader(projectPath)
    }

    static hasFacet() {
        return false // TODO: Need a `cds.requires.auditlog` or some other way to detect if enabled.
    }

    async run() {
        await this.runDependentMerging()
    }

    async runDependentMerging(context = {}) {
        const projectDescriptor = await this.projectReader.read(this.options)
        const { isNodejs, hasMta, hasHelm } = projectDescriptor.cap

        if (hasMta) {
            const mtaYAMLPath = join(this.projectPath, 'mta.yaml')
            const srv = isNodejs ? srvNode : srvJava
            await mergeYAML(
                mtaYAMLPath,
                join(__dirname, 'files', 'mta.yaml.hbs'),
                projectDescriptor,
                {
                    additions: [srv, auditlog],
                    relationships: [{
                        insert: [auditlog, 'name'],
                        into: [srv, 'requires', 'name']
                    }]
                }
            )
        }

        if (hasHelm) {
            //in case facet is being added to the already exisiting charts folder
            let isIndependentCommand = false
            if(Object.keys(context).length == 0){
                isIndependentCommand = true
                context.projectPath = this.projectPath
                context.oldTrackingData = context.newTrackingData = await parseMd5File(join(this.projectPath, 'chart', '.cds-add-helm-files.md5'))
            }

            await mergeYAML(
                join(this.projectPath, 'chart', 'values.yaml'),
                {
                    srv: { bindings: {
                        auditlog: { serviceInstanceName: 'auditlog' }
                    }},
                    auditlog: {
                        serviceOfferingName: 'auditlog-api',
                        servicePlanName: 'default'
                    }
                }
            )
            await copyAndTrack(join(__dirname, 'files', 'auditlog.yaml'), join(this.projectPath, 'chart', 'templates', 'auditlog.yaml'), context)
            if(isIndependentCommand) await writeMd5File(join(this.projectPath, 'chart', '.cds-add-helm-files.md5'), context.newTrackingData)
        }
    }
}
