// @sap/cds-dk/lib/init/template/destinations/index.js
const { join } = require('path')
const ProjectReader = require('../../util/projectReader')
const { mergeYAML } = require('../../util/merge')
const { copyAndTrack, parseMd5File, writeMd5File } = require('../../util/md5Tracking')

module.exports = class DestinationsTemplate extends require('../templateBase') {
    constructor(projectPath, generator) {
        super(projectPath, generator, __dirname)
        this.projectReader = new ProjectReader(projectPath)
    }

    async canRun() {
        const projectDescriptor = await this.projectReader.read(this.options)
        const { hasMta, hasHelm } = projectDescriptor.cap
        if (!hasHelm && hasMta) {
            throw `'cds add destinations' is not available for Cloud Foundry yet`
        }
        return true
    }

    static hasFacet(env) {
        const kinds = { odata: 1, 'odata-v2': 1, 'odata-v4': 1, rest: 1 }
        return Object.entries(env.requires).some(([service, require]) => !kinds[service] && kinds[require.kind])
    }

    async run() {
        await this.runDependentMerging()
    }

    async runDependentMerging(context = {}) {
        const projectDescriptor = await this.projectReader.read(this.options)
        const { hasHelm, hasDestination, hasHtml5Repo } = projectDescriptor.cap

        if (hasHelm) {
            //const env = await this.getEnv()
            const binding = hasDestination && { srv: { bindings: { destination: { serviceInstanceName: 'destinations' }}}}
            const html5Binding = hasHtml5Repo && { 'html5-apps-deployer': { bindings: { destination: { serviceInstanceName: 'destination' }}}}

            //in case facet is being added to the already exisiting charts folder
            let isIndependentCommand = false
            if(Object.keys(context).length == 0){
                isIndependentCommand = true
                context.projectPath = this.projectPath
                context.oldTrackingData = context.newTrackingData = await parseMd5File(join(this.projectPath, 'chart', '.cds-add-helm-files.md5'))
            }
            await mergeYAML(
                join(this.projectPath, 'chart', 'values.yaml'),
                {
                    ...binding,
                    ...html5Binding,
                    destination: {
                        serviceOfferingName: 'destination',
                        servicePlanName: 'lite',
                        parameters: {
                            version: '1.0.0',
                            ...(hasHtml5Repo) && { HTML5Runtime_enabled: true }
                        }
                    }
                }
            )
            await copyAndTrack(join(__dirname, 'files', 'destination.yaml'), join(this.projectPath, 'chart', 'templates', 'destination.yaml'), context)
            if(isIndependentCommand) await writeMd5File(join(this.projectPath, 'chart', '.cds-add-helm-files.md5'), context.newTrackingData)
        }
    }
}
